### ISSUE TITLE
API Proposal: WithDockerfile, WithBuildArg, and WithSecretBuildArg

### ISSUE THREAD

**Issue by mitchdenny** [dotnet/aspire#4346]
## Background and Motivation

Developers have requested the ability for Aspire to build a container image on the fly from a Dockerfile (see: https://github.com/dotnet/aspire/issues/1852). DCP added the capability to do this so this change is about adding this feature into the app model.

Current PR: https://github.com/dotnet/aspire/pull/4312

## Proposed API

```diff
+ public IResourceBuilder<ContainerResource> WithDockerfile(
+    this IResourceBuilder<ContainerResource> builder,
+    string contextPath,
+    string? dockerfilePath = null,
+    string? stage = null)

+    public static IResourceBuilder<T> WithBuildArg<T>(this IResourceBuilder<T> builder, string name, object value) where T : ContainerResource
+    public static IResourceBuilder<T> WithBuildArg<T>(this IResourceBuilder<T> builder, string name, IResourceBuilder<ParameterResource> value) where T : ContainerResource
+    public static IResourceBuilder<T> WithSecretBuildArg<T>(this IResourceBuilder<T> builder, string name, IResourceBuilder<ParameterResource> value) where T : ContainerResource

+ internal class DockerfileBuildAnnotation(string contextPath, string dockerfilePath, string? stage) : IResourceAnnotation
+ {
+     public string ContextPath => contextPath;
+     public string DockerfilePath = dockerfilePath;
+     public string? Stage => stage;
+     public Dictionary<string, object> BuildArguments { get; } = new();
+     public Dictionary<string, object> SecretBuildArguments { get; } = new();
+ }
```

### Manifest output

```json
    // snip
    "mycontainer": {
      "type": "container.v0",
      "build": {
        "context": "qots",
        "dockerfile": "qots/Dockerfile",
        "args": {
          "GO_VERSION": "{goversion.value}"
        },
        "secrets": {
          "MY_SECRET": "{mysecret.value}"
        }
      }
    }
    // snip
```

(update: if the docker build annotation is set, the image field is omitted. The `image` and `build` fields are exclusive in `container.v1`.

## Usage Examples

```csharp
// Add a container which builds from a Dockerfile given a context path.
var container = builder.AddContainer("mycontainer", "myimage")
                       .FromDockerfile("containers/mycustomcontainer");

// Add a container which builds from a Dockerfile given a context path and a custom Dockerfile name.
var container = builder.AddContainer("mycontainer", "myimage")
                       .FromDockerfile("containers/mycustomcontainer", "Otherdockerfile");

// Add a container which builds from a Dockerfile given a context path and a custom Dockerfile name.
var container = builder.AddContainer("mycontainer", "myimage")
                       .FromDockerfile("containers/mycustomcontainer")
                       .WithBuildArg("SOMEOPTION", 1);

// ... using parameters
var someOption = builder.AddParameter("someOption");
var container = builder.AddContainer("mycontainer", "myimage")
                       .FromDockerfile("containers/mycustomcontainer")
                       .WithBuildArg("SOMEOPTION", someOption);

// ... specifying build stage explicitly
var container = builder.AddContainer("mycontainer", "myimage")
                       .FromDockerfile("containers/mycustomcontainer", stage: "intermediate");
```

## Alternative Designs

We could have implemented a completely custom resource `AddDockerfile` but this would have meant quite a bit of duplication of extension methods was necessary. Additionally a container image that is built from a Dockerfile is conceptually identical in Aspire other than the fact that the image is built on the fly via DCP ... so this option just adds an annotation to capture the extra detail we need.

## Risks

We need to consider what would happen if a deployment tool read the manifest and didn't understand the build arguments and tried to pull the image from a public registry. Could this be an attack vector?

To help mitigate this, should we ... if the build metadata is present change the resource type to dockerfile.v1?

**Comment by danegsta** [comment]
There can be some noticeable differences between how a user might run a container locally for development vs. how they would build and run a container for production. A common pattern for development in many languages is to volume mount your local source into the container to enable the equivalent of hot reload; the Dockerfiles scaffolded by VS and VS Code for .NET projects actually include a stage for development that skips copying files into the container to speed up startup. So the volume and stage settings used for development may not match those needed for production.

**Comment by eerhardt** [comment]
1. How do I set the build directory?
2. Is "Dockerfile" the best name here, even if I'm using podman?

**Comment by danegsta** [comment]
For 1, the first argument to FromDockerfile is a path to the build directory.
For 2, it’s messy, because Dockerfile has basically become the industry standard and is the only file name you can use if you want full compatibility between Docker and Podman when building an image (.dockerignore vs .containerignore being another similar case).

**Comment by afscrome** [comment]
How would you pass secrets to the docker build with this api.  For example to `dotnet restore` from an Azure devops feed:

```ps
$env:AdoToken = az account get-access-token --resource 499b84ac-1321-427f-aa17-267ca6975798 --query accessToken --output tsv
docker build . --secret id=ado-accesstoken,env=AdoToken
```
With dockerfile:
```dockerfile

RUN --mount=type=secret,id=ado-accesstoken \
    export VSS_NUGET_ACCESSTOKEN=$(cat /run/secrets/ado-accesstoken); \
    dotnet restore
```

**Comment by eerhardt** [comment]
Is the `dockerfilePath` relative to the `contextPath`? Or relative to the AppHost directory? On the command line it is relative to the CWD, which in this case is the AppHost directory.

**Comment by danegsta** [comment]
> How would you pass secrets to the docker build with this api. For example to `dotnet restore` from an Azure devops feed:
> 
> ```postscript
> $env:AdoToken = az account get-access-token --resource 499b84ac-1321-427f-aa17-267ca6975798 --query accessToken --output tsv
> docker build . --secret id=ado-accesstoken,env=AdoToken
> ```
> 
> With dockerfile:
> 
> ```dockerfile
> RUN --mount=type=secret,id=ado-accesstoken \
>     export VSS_NUGET_ACCESSTOKEN=$(cat /run/secrets/ado-accesstoken); \
>     dotnet restore
> ```

For build time secrets, --build-arg parameters are treated as environment variables for build commands, but aren't persisted in the final image, so that's one option for passing a secret like an azdo token into a container for build time restore. Otherwise, using `RUN --mount=type=secret` in a Dockerfile is a way to make specific files available as temporary secrets during build.

**Comment by danegsta** [comment]
In which case, yeah, we'd need to include `--secret` arguments for buildx to handle the RUN mount. Good point.

**Comment by afscrome** [comment]
> For build time secrets, `--build-arg` parameters ... aren't persisted in the final image, so that's one option for passing a secret like an azdo token into a container for build time restore. 

Build Args do end up in the final image - run `docker history` on an image and it'll show you the build args in clear text.  (If you have a multi stage build, it's possible you've accidently mitigated the issue by copying from a previous layer that used a build arg rather than building on top of that layer)

**Comment by danegsta** [comment]
> > For build time secrets, `--build-arg` parameters ... aren't persisted in the final image, so that's one option for passing a secret like an azdo token into a container for build time restore.
> 
> Build Args do end up in the final image - run `docker history` on an image and it'll show you the build args in clear text. (If you have a multi stage build, it's possible you've accidently mitigated the issue by copying from a previous layer that used a build arg rather than building on top of that layer)

If you have an ARG statement in your Dockerfile, `docker history` will show that and the default value, but it doesn't include overrides passed in via the command line unless they were included verbatim as a command line argument (used as an argument in a  `RUN` command, for example, rather than as an ambient environment variable, but that's a problem you could encounter with secrets as well).

**Comment by DamianEdwards** [comment]
This is proposing a new builder method prefix, `From`. We should also consider applicability of existing terms or whether a different new term would be a better fit:

- `WithDockerfile`
    Pretty good alignment with existing methods that impact the resource like `WithEnvironment`, `WithLaunchProfile`, etc., and we have examples of `With` methods for container resources that impact the creation of the resource already, e.g. `WithImage`.
- `AsDockerfile`:
    The `As` prefix seems slightly more aligned with what the resource produces rather than where it comes from, `RunAsEmulator`, but they're pretty close
- `AddDockerfile`
    The `Add` prefix on an `IResourceBuilder<T>` represents adding a child resource, e.g. `AddDatabase` so doesn't seem like a good fit here
- `UseDockerfile`
    We don't currently use the `Use` prefix on `IResourceBuilder<T>` that I could find but it seems it would be an OK fit for this case but it could be argued it's basically the same as `With`

Based on above we should probably consider using `WithDockerfile` unless we feel strongly that this method needs to be differentiated as it's mutually exclusive to methods like `WithImage`/`WithImageTag`/`WithRegistry` for creating a container resource (is it?).

**Comment by mitchdenny** [comment]
> This is proposing a new builder method prefix...

Yeah, I was not sure on the name myself (note the name of the branch on the PR :)). I can easily be persuaded to have WithDockerfile. I'm not sold on `AsDockerfile` and in fact I think we misnamed this for `PublishAsDockerfile` (probably should have been `PublishAsContainer`. `AddX` and `UseX` are used in different contexts.

**Comment by mitchdenny** [comment]
> There can be some noticeable differences between how a user might run a container locally for development vs. how they would build and run a container for production...

Very true, although I'm not sure that means we need any API changes here. People can use if/switch/tenary operators to provide customization to the way the container is configured based on whether they are publishing the manifest or any other value.

What we provide here is something they can easily compose into their final solution.

**Comment by mitchdenny** [comment]
> How would you pass secrets to the docker build with this api ....

This isn't something I've modeled with this. I could potentially add something like `WithSecretBuildArg(...)`. But before we could tackle that we would need DCP to support it (@danegsta @karolz-ms?)

Locally I think this could be pretty straight forward - but for manifest publishing scenarios we would need to think how this is going to work. We would probably want to make `WithSecretBuildArg(...)` only accept an `IResourceBuilder<ParameterResource>` to ensure that all the secrets are externalized.

**Comment by mitchdenny** [comment]
> If you have an ARG statement in your Dockerfile, `docker history` will show that and the default value, but it doesn't include overrides passed in via the command line unless they were included verbatim as a command line argument (used as an argument in a `RUN` command, for example, rather than as an ambient environment variable, but that's a problem you could encounter with secrets as well).

We definitely want to handle secrets with care. I'm thinking that we do a few things here.

1. We add `IResourceBuilder<> WithSecretBuildArg<T>(this IResourceBuilder<T>, IResourceBuilder<ParameterResource>) where T: ContainerResource
2. We make `WithBuildArgs` throw if the parameter passed in is a secret (with instructions to use WithSecretBuildArg instead.

Thoughts? This is obviously predicated on DCP supporting it - which I think is a good idea.

**Comment by danegsta** [comment]
> > How would you pass secrets to the docker build with this api ....
> 
> This isn't something I've modeled with this. I could potentially add something like `WithSecretBuildArg(...)`. But before we could tackle that we would need DCP to support it (@danegsta @karolz-ms?)
> 
> Locally I think this could be pretty straight forward - but for manifest publishing scenarios we would need to think how this is going to work. We would probably want to make `WithSecretBuildArg(...)` only accept an `IResourceBuilder<ParameterResource>` to ensure that all the secrets are externalized.

I opened up a DCP PR to add secrets support and use the image ID as the run source, hope to have it make its way upstream early next week.

**Comment by mitchdenny** [comment]
Awesome. Once that lands in the repo we can get the secrets working end to end in my PR.

**Comment by danegsta** [comment]
> > If you have an ARG statement in your Dockerfile, `docker history` will show that and the default value, but it doesn't include overrides passed in via the command line unless they were included verbatim as a command line argument (used as an argument in a `RUN` command, for example, rather than as an ambient environment variable, but that's a problem you could encounter with secrets as well).
> 
> We definitely want to handle secrets with care. I'm thinking that we do a few things here.
> 
> 1. We add `IResourceBuilder<> WithSecretBuildArg(this IResourceBuilder, IResourceBuilder) where T: ContainerResource
> 2. We make `WithBuildArgs` throw if the parameter passed in is a secret (with instructions to use WithSecretBuildArg instead.
> 
> Thoughts? This is obviously predicated on DCP supporting it - which I think is a good idea.

Technically build args are the only way to pass ephemeral secrets in; the secrets argument is for mounting secret files at build time.

**Comment by mitchdenny** [comment]
> Is the `dockerfilePath` relative to the `contextPath`? Or relative to the AppHost directory? On the command line it is relative to the context directory.

The way it works in the current implementation is this:

1. The contextPath is relative to the app host directory ... unless it is absolute.
2. The dockerfilePath is relative to the contextPath ... unless it is absolute.
3. When the values are handed off to DCP they are made absolute.
4. When the values are emitted into the manifest, they are relative to the manifest file location.

**Comment by mitchdenny** [comment]
> Technically build args are the only way to pass ephemeral secrets in; the secrets argument is for mounting secret files at build time.

Good to know. Is this behavior consistent across Docker and Podman?

**Comment by mitchdenny** [comment]
@ellismg @vhvb1989 @prom3theu5 @davidfowl 

Wanted to call out a specific risk I saw here. Lets say I have the following code:

```csharp
builder.AddContainer("mycontainer", "myinternalimage").WithDockerfile(...)
```

And then lets say I run a version of AZD or Aspir8 that does not support the new `build`: field and skips over it. Then it's going to by default let the container runtime resolve `myinternalimage` from its default registry.

Potentially a malicious actor could learn of this image name and create a repository that matches in that public registry and whereby inject malicious code into the developers app (possibly in a context where production secrets are exposed to the container).

Doing a schema revision is not sufficient here because current deployment tools do not look at the schema version. In this scenario I think we might have a case for changing the resource version to `container.v1`.

Thoughts?

**Comment by afscrome** [comment]
`WithSecretBuildArg` as a name seems a bit ambiguous - are we talking about secrets, are we talking about build args, or are we talking about a build arg that is thought of as a secret?  I'd prefer to see two separate methods - `WithBuildArg` (`--build-arg`) and `WithSecret` (`--secret`) (or possibly `WithBuildSecret`).

Agree that `ParameterResource` on build args shouldn't be blocked outright though `--build-arg` as are sometimes unavoidable depending on how the image, or if you're using tools purely for local dev, maybe you don't consider the values as a secret but the Aspire APIs force you to plumb them through `ParameterResource`.

That said, could still encourage the use of secrets over build args by using `[Obsolete]` attribute or roslyn analyser to warn and recommend the use of secrets if passing a ` ParameterResource` as a build arg.

**Comment by danegsta** [comment]
> > Technically build args are the only way to pass ephemeral secrets in; the secrets argument is for mounting secret files at build time.
> 
> Good to know. Is this behavior consistent across Docker and Podman?

Yeah, the build secrets argument has the same behavior between Docker and Podman.

**Comment by danegsta** [comment]
Theoretically registry namespaces prevent a user from squatting on an arbitrary image. Single name registries such as `node` or `postgres` can only be created by Docker, Inc. and any given registry prefix (such as `dpage/` for `dpage/pgadmin4`) is owned by a specific user/organization. I think the only real attack vectors would be a compromised registry/repository, an image reference to an existing malicious registry, or a reference to a non-existent registry prefix.

**Comment by davidfowl** [comment]
> And then lets say I run a version of AZD or Aspir8 that does not support the new build: field and skips over it. Then it's going to by default let the container runtime resolve myinternalimage from its default registry.

This is a reason to change the version of the resource. If these properties are added to the manifest it's container.v1, if the v0 compatible properties are added it's v0.

**Comment by mitchdenny** [comment]
@davidfowl yeah that is what I'm thinking.

@danegsta you are probably right about this being an unlikely threat given the way current major registries work. But I'm wondering whether we should still mitigate the threat instead of relying on that.

**Comment by danegsta** [comment]
@mitchdenny DCP support for build secrets has been merged into main.

**Comment by afscrome** [comment]
When building dockerfiles, should there be any default labels applied to the images - e.g. [open container](https://github.com/opencontainers/image-spec/blob/main/annotations.md) labels, or any aspire specific labels to describe where they came from (e.g. resource name, project name)

**Comment by mitchdenny** [comment]
@afscrome this change has just merged in but I think your suggestion above is worth considering. Did you want to open up a separate issue?
